server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -116,22 +120,35 @@ public final class IndicesRequestCache implements RemovalListener<IndicesRequest
     private final Set<CleanupKey> keysToClean = ConcurrentCollections.newConcurrentSet();
     private final ByteSizeValue size;
     private final TimeValue expire;
-    private final Cache<Key, BytesReference> cache;
     private final Function<ShardId, Optional<CacheEntity>> cacheEntityLookup;
+    private final ICache<Key, BytesReference> cache;
 
     IndicesRequestCache(Settings settings, Function<ShardId, Optional<CacheEntity>> cacheEntityFunction) {
         this.size = INDICES_CACHE_QUERY_SIZE.get(settings);
         this.expire = INDICES_CACHE_QUERY_EXPIRE.exists(settings) ? INDICES_CACHE_QUERY_EXPIRE.get(settings) : null;
         long sizeInBytes = size.getBytes();
-        CacheBuilder<Key, BytesReference> cacheBuilder = CacheBuilder.<Key, BytesReference>builder()
-            .setMaximumWeight(sizeInBytes)
-            .weigher((k, v) -> k.ramBytesUsed() + v.ramBytesUsed())
-            .removalListener(this);
+        this.cacheEntityLookup = cacheEntityFunction;
+        String cacheTypeInString = FeatureFlags.INDICES_REQUEST_CACHE_TYPE.get(settings);
+        StoreAwareCacheBuilder<Key, BytesReference> openSearchOnHeapCacheBuilder = new OpenSearchOnHeapCache.Builder<Key, BytesReference>()
+            .setMaximumWeightInBytes(sizeInBytes)
+            .setWeigher((k, v) -> k.ramBytesUsed() + v.ramBytesUsed());
         if (expire != null) {
-            cacheBuilder.setExpireAfterAccess(expire);
+            openSearchOnHeapCacheBuilder.setExpireAfterAccess(expire);
+        }
+        CacheType cacheType;
+        try {
+            cacheType = CacheType.valueOf(cacheTypeInString);
+        } catch (IllegalArgumentException ex) {
+            cacheType = CacheType.ON_HEAP; // Default to onHeap
+        }
+        switch (cacheType) {
+            case ON_HEAP:
+                openSearchOnHeapCacheBuilder.setEventListener(this);
+                cache = openSearchOnHeapCacheBuilder.build();
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown cache type");
```

Comment:
Don't want to default to ON_HEAP here?

Reply:
And could log more details about what unknown cache type was passed in to debug if it ever comes to this